### ISSUE TITLE
Recursive event loop

### ISSUE THREAD

**Issue by antoyo** [antoyo/futures-glib-rs#2]
@alexcrichton 
Hi, I found out that nested event loop does not actually work with `futures-glib`.

So, here's what I tried:
I made the source recursive:
```rust
source.set_can_recurse(true);
```
so that the `dispatch` function can be called recursively.

The problem is that a `BorrowMut` error will show up because the `dispatch` function will be called before the end of the call to `poll_future_notify` so that the task is still borrowed when trying to borrow it again in the recursive call.

Do you have an idea on how I could allow the nested event loop?
In other words, is there a way to poll a future while this specific future is currently being pulled? (**edit:** Oh, I think it does not make sense: even if I was able to do that, I'll end up with two mutable references which is not sound)

I can show you a code that triggers this issue, but it is not a simple example, it is in my relm crate, so it's not easy to figure out what happens.

Thanks for your help.

**Edit: ** As far as I understand, it is not correct to support recursive event loop. Am I right?

**Comment by alexcrichton** [comment]
Oh I personally thinking that nested event loops are fine, a `BorrowMut` error though probably just means that a `RefCell` is being held for a little bit too long. Do you have an test case I could poke around with locally?

**Comment by antoyo** [comment]
I work on a branch of `relm` and when it's merged I'll show you how to reproduce this issue.
That might take a couple of weeks until it's ready, thought.
Anyway, thanks for your help.

**Comment by alexcrichton** [comment]
Ok sounds good!

**Comment by antoyo** [comment]
Finally, I don't think we should support that.
I think it doesn't make sense because `Stream::poll()` takes `self` by a mutable reference, so `poll()` cannot be called recursively, even indirectly.

One use case for that would be that the second call to `poll()` is made on another `Stream`, but I believe it is not worth supporting that.
